File: 백준/Gold/1939. 중량제한/중량제한.py
from collections import deque

def bfs(graph, start, end, weight_limit, n):
    queue = deque([start])
    visited = [False] * (n + 1)
    visited[start] = True
    
    while queue:
        current = queue.popleft()
        
        if current == end:
            return True
        
        for neighbor, weight in graph[current]:
            if not visited[neighbor] and weight >= weight_limit:
                visited[neighbor] = True
                queue.append(neighbor)
    
    return False

def find_max_weight(n, bridges, start, end):
    graph = [[] for _ in range(n + 1)]
    for u, v, w in bridges:
        graph[u].append((v, w))
        graph[v].append((u, w))
    
    low, high = 1, max(w for _, _, w in bridges)
    
    while low <= high:
        mid = (low + high) // 2
        if bfs(graph, start, end, mid, n):
            low = mid + 1
        else:
            high = mid - 1
    
    return high

import sys
input = sys.stdin.read
data = input().split()

n, m = int(data[0]), int(data[1])
bridges = [(int(data[i]), int(data[i+1]), int(data[i+2])) for i in range(2, 3*m+2, 3)]
start_island, end_island = int(data[-2]), int(data[-1])

print(find_max_weight(n, bridges, start_island, end_island))

File: 백준/Gold/1939. 중량제한/중량제한.java
import java.io.*;
import java.util.*;

public class Main {
    static ArrayList<ArrayList<Bridge>> graph;
    
    static class Bridge {
        int destination;
        int weight;
        
        public Bridge(int destination, int weight) {
            this.destination = destination;
            this.weight = weight;
        }
    }
    
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());
        
        int n = Integer.parseInt(st.nextToken());
        int m = Integer.parseInt(st.nextToken());
        
        // 그래프 초기화
        graph = new ArrayList<>();
        for (int i = 0; i <= n; i++) {
            graph.add(new ArrayList<>());
        }
        
        // 다리 정보 저장 및 최대 무게 찾기
        int maxBridgeWeight = 0;
        for (int i = 0; i < m; i++) {
            st = new StringTokenizer(br.readLine());
            int u = Integer.parseInt(st.nextToken());
            int v = Integer.parseInt(st.nextToken());
            int w = Integer.parseInt(st.nextToken());
            
            graph.get(u).add(new Bridge(v, w));
            graph.get(v).add(new Bridge(u, w));
            
            maxBridgeWeight = Math.max(maxBridgeWeight, w);
        }
        
        st = new StringTokenizer(br.readLine());
        int startIsland = Integer.parseInt(st.nextToken());
        int endIsland = Integer.parseInt(st.nextToken());
        
        // 이분 탐색으로 최대 무게 찾기
        int result = findMaxWeight(n, startIsland, endIsland, maxBridgeWeight);
        System.out.println(result);
        
        br.close();
    }
    
    static boolean bfs(int start, int end, int weightLimit, int n) {
        Queue<Integer> queue = new LinkedList<>();
        boolean[] visited = new boolean[n + 1];
        
        queue.offer(start);
        visited[start] = true;
        
        while (!queue.isEmpty()) {
            int current = queue.poll();
            
            if (current == end) {
                return true;
            }
            
            for (Bridge bridge : graph.get(current)) {
                if (!visited[bridge.destination] && bridge.weight >= weightLimit) {
                    visited[bridge.destination] = true;
                    queue.offer(bridge.destination);
                }
            }
        }
        
        return false;
    }
    
    static int findMaxWeight(int n, int start, int end, int maxBridgeWeight) {
        int low = 1;
        int high = maxBridgeWeight;
        
        while (low <= high) {
            int mid = (low + high) / 2;
            if (bfs(start, end, mid, n)) {
                low = mid + 1;
            } else {
                high = mid - 1;
            }
        }
        
        return high;
    }
}